src/wrapper/clap.rs

```diff
@@ -44,7 +44,7 @@ macro_rules! nih_export_clap {
                     && unsafe { ::std::ffi::CStr::from_ptr(factory_id) }
                         == ::nih_plug::wrapper::clap::CLAP_PLUGIN_FACTORY_ID
                 {
-                    &(*FACTORY).clap_plugin_factory as *const _ as *const ::std::ffi::c_void
+                    &*FACTORY as *const _ as *const ::std::ffi::c_void
```

Comment:
This technically works thanks to `repr(C)`, but it's also unsound. You're supposed to return a `const clap_plugin_factory_t*` here and now this returns a `const nih_plug::wrapper::clap::Factory<$plugin_ty>*`. Did miri complain about this pointer being used inside of NIH-plug's factory implementation? That's probably the place where it should be modified instead.

Reply:
That's actually the other way around: casting from `nih_plug::wrapper::clap::Factory<$plugin_ty>*`  to `clap_plugin_factory_t*`  *is* sound thanks to `repr(C)`, which guarantees the order of fields, and that the two raw pointers are equivalent, matching C's behavior (cf. [the nomicon](https://doc.rust-lang.org/nomicon/other-reprs.html#reprc), which explicitly mentions the C-style reinterpreting of values as different types as an use-case for `repr(C)`).

Here's the full MIRI output, for reference (I should probably have included it in my PR, my bad):
```test it_works ... error: Undefined Behavior: trying to retag from <278063> for SharedReadOnly permission at alloc1[0x18], but that tag does not exist in the borrow stack for this location
   --> /home/adrien/Projects/RustAudio/nih-plug-miri/src/wrapper/clap/factory.rs:47:23
    |
47  |         let factory = &*(factory as *const Self);
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                       |
    |                       trying to retag from <278063> for SharedReadOnly permission at alloc1[0x18], but that tag does not exist in the borrow stack for this location
    |                       this error occurs as part of retag at alloc1[0x0..0x128]
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <278063> was created by a SharedReadOnly retag at offsets [0x0..0x18]
   --> host/tests/nih-plug-loading.rs:208:1
    |
208 | nih_export_clap!(Gain);
    | ^^^^^^^^^^^^^^^^^^^^^^
```

MIRI does complain about the usage in the factory implementation, but it points at the retag (i.e. reborrow) made in the entry generated by the macro. The issue isn't with where the reference points to (it's the same in either case thanks to `repr(C)`), but the region of memory that is being borrowed (which is how `rustc`, and therefore MIRI, see it).

When writing `&(*FACTORY).clap_plugin_factory`, what it's doing is taking the borrow of the whole `Factory<$plugin_ty>` (here `alloc1[0x0..0x128]`), and creates a new reference to the `clap_plugin_factory_t` subfield (`alloc1[0x0..0x18]`), which is then being cast to the returned raw pointer.

That reference *only* allows access to the `clap_plugin_factory` part, and not the rest of the struct after it, because since you're re-borrowing only a subfield, the other subfield of that struct could still be borrowed elsewhere. (Although I don't believe `rustc` currently makes anything with that assumption, which is probably why it can't miscompile yet). That's why MIRI complains as soon as you're trying to read past that region, e.g. `alloc1[0x18]`.

That's also why changing it in the factory implementation wouldn't work: there's no way to get the full range of the `Factory<$plugin_ty>` from that reference, because it's the entry implementation that overly restricts it.

Reply:
I was thinking in terms of C UB (like I explained on the Discord), but this specific cast between pointers to struct `foo` and pointers to struct `bar` when `foo`'s first field is a `bar` is apparently specifically specified as valid in the C standard. I somehow assumed it was one of those things everyone (including myself) does but that isn't actually technically valid. An alternative approach here would have been to move some more logic from the generic factory struct to macros so the macro can be generated statically. But that gets a lot uglier.